### PR TITLE
PlotReport causes error in MNIST example when matplotlib doesn't exist

### LINKED ISSUE #2276: Chainer 1.21.0: mnist example raises AttributeError: 'PlotReport' object has no attribute '_file_name'
I ran an example code train_mnist.py without matplotlib and got AttributeError.

1. create Anaconda environment
$ conda create -n testenv python=2.7

```
Fetching package metadata .......
Solving package specifications: ..........
Using Anaconda Cloud api site https://api.anaconda.org

The following NEW packages will be INSTALLED:

    openssl:    1.0.2k-0
    pip:        9.0.1-py27_1
    python:     2.7.13-0
    readline:   6.2-2
    setuptools: 27.2.0-py27_0
    sqlite:     3.13.0-0
    tk:         8.5.18-0
    wheel:      0.29.0-py27_0
    zlib:       1.2.8-3

```
2. install chainer with pip
$ pip install chainer

3. clone Chainer repository
$ git clone https://github.com/pfnet/chainer.git
$ cd chainer
$ git log

```
commit 6d2855c85c86cf655e7b87991ac353fa9f8d7603
Merge: 5f18ac0 10fb6ed
Author: Ryosuke Okuta <okuta@preferred.jp>
Date:   Wed Feb 15 10:18:25 2017 +0900

    Merge pull request #2273 from yuyu2172/missing-docs

    Include space2depth and depth2space to docs

```

4. run mnist example
$ cd examples/mnist/
$ python train_mnist.py -g 0

5. got error

```
GPU: 0
# unit: 1000
# Minibatch-size: 100
# epoch: 20

chainer/training/extensions/plot_report.py:27: UserWarning: matplotlib is not installed on your environment, so nothing will be plotted at this time. Please install matplotlib to plot figures.

  $ pip install matplotlib

  warnings.warn('matplotlib is not installed on your environment, '

Traceback (most recent call last):
  File "train_mnist.py", line 115, in <module>
    main()
  File "train_mnist.py", line 112, in main
    trainer.run()
  File "<secret>/python2.7/site-packages/chainer/training/trainer.py", line 298, in run
    entry.extension(self)
  File "<secret>/python2.7/site-packages/chainer/training/extensions/_snapshot.py", line 78, in snapshot
    _snapshot_object(trainer, trainer, filename.format(trainer), savefun)
  File "<secret>/python2.7/site-packages/chainer/training/extensions/_snapshot.py", line 88, in _snapshot_object
    savefun(tmppath, target)
  File "<secret>/python2.7/site-packages/chainer/serializers/npz.py", line 67, in save_npz
    s.save(obj)
  File "<secret>/python2.7/site-packages/chainer/serializer.py", line 69, in save
    obj.serialize(self)
  File "<secret>/python2.7/site-packages/chainer/training/trainer.py", line 318, in serialize
    entry.extension.serialize(s[name])
  File "<secret>/python2.7/site-packages/chainer/training/extensions/plot_report.py", line 150, in serialize
    serializer('_plot_{}'.format(self._file_name),
AttributeError: 'PlotReport' object has no attribute '_file_name'

```

### PR DESCRIPTION
**Pull request by mitmul**
Fix #2276

### CODE REVIEW

**Comment by delta2323** [comment]
In the offline discussion, there was an idea that we do not instantiate when `matplotlib` is not available. Could you consider its feasibility?

File: examples/mnist/train_mnist.py [hunk 1]
```diff
@@ -85,12 +85,14 @@ def main():
     trainer.extend(extensions.LogReport())
 
     # Save two plot images to the result dir
-    trainer.extend(
-        extensions.PlotReport(['main/loss', 'validation/main/loss'], 'epoch',
-                              file_name='loss.png'))
-    trainer.extend(
-        extensions.PlotReport(['main/accuracy', 'validation/main/accuracy'],
-                              'epoch', file_name='accuracy.png'))
+    if extensions.PlotReport.available:
```
Comment (ir5) [inline examples/mnist/train_mnist.py]: `extensions.PlotReport.available` is a function, thus this expression will be always evaluated as true. Is it reasonable to write this as a function-call form: `extensions.PlotReport.available()`?

File: chainer/training/extensions/plot_report.py [hunk 2]
```diff
@@ -92,6 +89,9 @@ def __init__(self, y_keys, x_key='iteration', trigger=(1, 'epoch'),
         self._init_summary()
         self._data = {k: [] for k in y_keys}
 
+    def available(self):
```
Comment (ir5) [inline chainer/training/extensions/plot_report.py]: `@classmethod` decorator seems to be required?

File: examples/mnist/train_mnist.py (hunk 1 above)
Comment (mitmul) [inline examples/mnist/train_mnist.py]: Oops, sorry it's a mistake. Another option is to make `available` a property of PlotReport. I'll make a modification based on this.

File: chainer/training/extensions/plot_report.py (hunk 2 above)
Comment (mitmul) [inline chainer/training/extensions/plot_report.py]: How about using `@property` here?

File: examples/mnist/train_mnist.py (hunk 1 above)
Comment (ir5) [inline examples/mnist/train_mnist.py]: Hmm, it seems that `@property` does not work on class methods. `extensions.PlotReport.available` is evaluated as `<property object at 0x*******>`.

Comment (mitmul) [inline examples/mnist/train_mnist.py]: Ah, right. I'll fix it by replacing it with `extensions.PlotReport.available()`

**Comment by ir5** [comment]
LGTM

File: chainer/training/extensions/plot_report.py [hunk 3]
```diff
@@ -92,6 +89,11 @@ def __init__(self, y_keys, x_key='iteration', trigger=(1, 'epoch'),
         self._init_summary()
         self._data = {k: [] for k in y_keys}
 
+    @classmethod
```
Comment (beam2d) [inline chainer/training/extensions/plot_report.py]: Is it intended to use `classmethod` here? If not, use `staticmethod` instead.

Comment (mitmul) [inline chainer/training/extensions/plot_report.py]: Sorry it's mistake. I'll fix it

**Comment by beam2d** [comment]
LGTM

**Comment by ISP-Tetsuro-Kitajima** [comment]
Thank you very much!!!

**Pull request merged**